File: web/src/api/services.ts
import { api } from './client'

export type ServiceStatus = 'running' | 'degraded' | 'stopped' | 'unknown'
export type ServiceSource = 'docker' | 'process' | 'manual'
export type LogSource = 'docker' | 'file' | 'none'

export interface ServiceInfo {
  id: number
  agent_id: number
  host: string
  name: string
  source: ServiceSource
  port: number
  metrics_url: string
  pid: number | null
  container_name: string | null
  model: string | null
  vllm_version: string | null
  started_at: string | null
  log_source: LogSource
  scrape_ok: boolean
  active: boolean
  last_seen_at: string | null
  status: ServiceStatus
}

export interface ServiceDetail extends ServiceInfo {
  cmdline: string | null
  cwd: string | null
  env: Record<string, string>
  container_id: string | null
  log_path: string | null
  profiler_dir: string | null
  first_seen_at: string
}

export type ActiveFilter = 'all' | 'true' | 'false'

export async function listServices(active: ActiveFilter = 'all'): Promise<ServiceInfo[]> {
  const res = await api.get<ServiceInfo[]>('/services', { params: { active } })
  return res.data
}

export async function getService(id: number): Promise<ServiceDetail> {
  const res = await api.get<ServiceDetail>(`/services/${id}`)
  return res.data
}

export async function listAgentServices(agentId: number): Promise<ServiceInfo[]> {
  const res = await api.get<ServiceInfo[]>(`/agents/${agentId}/services`)
  return res.data
}

/** Default ordering: stopped services last, everything else by host then name. */
export function sortServices(list: ServiceInfo[]): ServiceInfo[] {
  return [...list].sort((a, b) => {
    const sa = a.status === 'stopped' ? 1 : 0
    const sb = b.status === 'stopped' ? 1 : 0
    if (sa !== sb) return sa - sb
    return a.host.localeCompare(b.host) || a.name.localeCompare(b.name)
  })
}

File: agent/ai_monitor_agent/services.py
"""Service model shared by discovery, metrics proxy, log tailing and heartbeat."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any, Literal

Source = Literal["docker", "process", "manual"]
LogSource = Literal["docker", "file", "none"]


@dataclass
class ServiceInfo:
    name: str
    source: Source
    port: int
    metrics_url: str
    pid: int | None = None
    container_id: str | None = None
    container_name: str | None = None
    log_source: LogSource = "none"
    log_path: str | None = None
    profiler_dir: str | None = None
    model: str | None = None
    vllm_version: str | None = None
    started_at: float | None = None
    cmdline: str | None = None
    cwd: str | None = None
    env: dict[str, str] = field(default_factory=dict)
    scrape_ok: bool = False

    def to_dict(self) -> dict[str, Any]:
        """Plain dict matching the server's ``ServiceIn`` schema field-for-field."""
        return asdict(self)
